server/middleware/identityMiddleware.ts

```diff
@@ -1,70 +1,161 @@
 	updateManageUrl,
-} from './requestMiddleware';
+} from '@/server/middleware/requestMiddleware';
+import {
+	allIdapiCookiesSet,
+	OAuthAccessTokenCookieName,
+	oauthCookieOptions,
+	OAuthIdTokenCookieName,
+	performAuthorizationCodeFlow,
+	sanitizeReturnPath,
+	scopes,
+	setLocalStateFromIdTokenOrUserCookie,
+	verifyOAuthCookies,
+} from '@/server/oauth';
+import { getConfig as getOktaConfig } from '@/server/oktaConfig';
+import {
+	getScopeFromRequestPathOrEmptyString,
+	X_GU_ID_FORWARDED_SCOPE,
+} from '@/shared/identity';
+import { requiresSignin } from '@/shared/requiresSignin';
 
-interface RedirectResponseBody extends IdentityDetails {
-	signInStatus: string;
-	redirect?: {
-		url: string;
+declare const CYPRESS: string;
+
+const handleIdentityMiddlewareError = (err: Error, res: Response) => {
+	console.log('OAuth / Middleware / Error', err);
+	res.redirect('/maintenance');
+};
+
+export const withIdentity: (statusCodeOverride?: number) => RequestHandler =
+	(statusCodeOverride?: number) =>
+	async (req: Request, res: Response, next: NextFunction) => {
+		if (CYPRESS === 'SKIP_IDAPI') {
+			return next();
+		}
+
+		try {
+			const { useOkta } = await getOktaConfig();
+			if (useOkta) {
+				return authenticateWithOAuth(req, res, next);
+			} else {
+				return authenticateWithIdapi(statusCodeOverride)(
+					req,
+					res,
+					next,
+				);
+			}
+		} catch (err) {
+			handleIdentityMiddlewareError(err, res);
+		}
 	};
-}
 
-const containsSignInTokenQueryParameters = (
-	req: MockableExpressRequest,
-): boolean =>
-	signInTokenQueryParameterNames.some(
-		(name) => req.query[name] !== undefined,
-	);
+export const authenticateWithOAuth = async (
+	req: Request,
+	res: Response,
+	next: NextFunction,
+) => {
+	// Get the path of the current page and use it as our returnPath after the OAuth callback.
+	const returnPath = sanitizeReturnPath(req.originalUrl);
+
+	try {
+		if (requiresSignin(req.originalUrl)) {
+			// The route requires signin
+			/////////////////////////////////////////////////////////////////////////////////////
+			// If we have a GU_SO cookie, we've signed out recently, so we need to delete
+			// the access and ID tokens from the browser and perform the OAuth flow again.
+			if (req.cookies['GU_SO']) {
+				res.clearCookie(OAuthAccessTokenCookieName, oauthCookieOptions);
+				res.clearCookie(OAuthIdTokenCookieName, oauthCookieOptions);
+				return performAuthorizationCodeFlow(req, res, {
+					redirectUri: `https://manage.${conf.DOMAIN}/oauth/callback`,
+					scopes,
+					returnPath,
+				});
+			}
 
-const redirectOrCustomStatusCode = (
-	res: express.Response,
-	redirectURL: string,
-	statusCode?: number,
-) =>
-	statusCode
-		? res.status(statusCode).header('Location', redirectURL).send()
-		: res.redirect(redirectURL);
+			const { accessToken, idToken } = await verifyOAuthCookies(req);
+			const idapiCookies = allIdapiCookiesSet(req);
+			if (idapiCookies && accessToken && idToken) {
```

Comment:
maybe we could get rid of `idapiCookies` here ? 

it's actually a boolean not the cookies themselves, I would suggest renaming it to `areIdapiCookiesSet` or something, but if it's only used in one place we might us well inline the call to allIdapiCookies set in the condition:
```typescript
 if (allIdapiCookiesSet(req) & accessToken && idToken)
```

Reply:
Good idea!